packages/webapp/src/_app/eos/secret-publisher.ts

```diff
@@ -0,0 +1,174 @@
+import ecc, { PrivateKey, PublicKey } from "eosjs-ecc";
+const ec = require("elliptic").ec("secp256k1");
+const bn = require("bn.js");
+
+export const publishSecretToChain = async (
+    message: string,
+    publisherAccount: string,
+    recipientAccounts: string[],
+    info?: string
+) => {
+    const [publisherKey, ...recipientKeys] = await fetchRecipientKeys([
+        publisherAccount,
+        ...recipientAccounts,
+    ]);
+    console.info(publisherKey, recipientKeys);
+
+    // TODO: check publisherkey matches the current one in browser
+
+    const transientKeyPair = await generateRandomTransientKey();
+    console.info(transientKeyPair);
+
+    const keks = await ecdhRecipientsKeyEncriptionKeys(
+        [publisherKey, ...recipientKeys],
+        transientKeyPair.privateKey,
+        info
+    );
+    console.info(keks);
+
+    const sessionKey = await generateRandomSessionKey();
+    console.info(sessionKey);
+
+    const encryptedSessionKeys = await encryptSessionKeys(sessionKey, keks);
+    console.info(encryptedSessionKeys);
+
+    const encryptedMessage = await encryptMessage(sessionKey, message);
+
+    // TODO: prepare publish trx to eos chain with all the prepared material
+    // eden.publishelection(
+    //     publisher = logged_in_user
+    //     round_no = '2',
+    //     encrypted_link,
+    //   // for each recipient
+    //     [(encrypted_session_keyN, recipient_keyN, transient_ecc_pubkey)]
+    //   )
+    console.info({
+        encryptedSessionKeys,
+        recipientKeys,
+        transientKeyPair,
+        encryptedMessage,
+    });
+
+    return {};
+};
+
+const fetchRecipientKeys = async (accounts: string[]) => {
+    // TODO: real implementation
+    // fetch eden.keys_table( where account = each account item )
+
+    const mockedKeysLen = accounts.length;
+    const mockedKeys = [];
+    for (let i = 0; i < mockedKeysLen; i++) {
+        mockedKeys.push(
+            "EOS5wGY8RcG8PuGYTPU9QwmiEm9XReBkK4SGtSuS4f1oeX4TNGaHT"
+        );
+        // temp PK: 5KcMypBPGByXbeBphsVzyyzUvg31tWNKsgrfXs2MmanYgYf4gao
+    }
+    return mockedKeys;
+};
+
+const generateRandomTransientKey = async () => {
+    const privateKey = await ecc.randomKey();
+    const publicKey = ecc.privateToPublic(privateKey);
+    return { privateKey, publicKey };
+};
+
+const ecdhRecipientsKeyEncriptionKeys = async (
+    recipientPublicKeys: string[],
+    transientPrivateKeyString: string,
+    info?: string
+): Promise<CryptoKey[]> => {
+    const transientPrivateKey = PrivateKey.fromString(
+        transientPrivateKeyString
+    );
+    const transientEc = ec.keyPair({
+        priv: new bn(transientPrivateKey.d.toHex(), 16),
+    });
+    return Promise.all(
+        recipientPublicKeys.map(async (publicKey: string) => {
+            const recipientEcPublicKey = ec.keyFromPublic(
+                PublicKey.fromString(publicKey).toHex(),
```

Comment:
What will happen if publicKey uses the P-256 curve instead of secp256k1?

Reply:
nice point. Do we need to enforce that all recipients have K1? what if recipients have different types of public keys? I think for now we should drive the type through the UI and generate the encryption keys as K1!

Reply:
The new code depends on all key types being the same:

```
const deriveEcdhSecret = (
    eosPrivateKeyA: PrivateKey,
    eosPublicKeyB: PublicKey
) => {
    return eosPrivateKeyA
        .toElliptic()
        .derive(eosPublicKeyB.toElliptic().getPublic())
        .toArrayLike(ArrayBuffer); // TODO: review endianness, empty = Big Endian
};
```

This will work if both keys are `k1` or both are `r1`.